[IsDeprecatedWeakRefSmartPointerException] Make WebDataListSuggestionPicker ref-counted

RupinMittal opened the pull request:
#### c2dea7aaf6eb365fe0ff325254c1a44aba2e6572
<pre>
[IsDeprecatedWeakRefSmartPointerException] Make WebDataListSuggestionPicker ref-counted
<a href="https://bugs.webkit.org/show_bug.cgi?id=281341">https://bugs.webkit.org/show_bug.cgi?id=281341</a>
<a href="https://rdar.apple.com/137771975">rdar://137771975</a>

Reviewed by Chris Dumez.

Remove IsDeprecatedWeakRefSmartPointerException] by making WebDataListSuggestionPicker
ref-counted. Since WebDataListSuggestionPicker could now outlive WebPage or
DataListSuggestionsClient, we make those members WeakPtr and CheckedPtr and null-check
them before use. Accordingly, we ensure the DataListSuggestionsClient is set to nullptr
before the DataListSuggestionsClient is destroyed.

* Source/WebCore/html/TextFieldInputType.cpp:
(WebCore::TextFieldInputType::~TextFieldInputType):
(WebCore::TextFieldInputType::handleKeydownEvent):
(WebCore::TextFieldInputType::didCloseSuggestions):
(WebCore::TextFieldInputType::displaySuggestions):
(WebCore::TextFieldInputType::closeSuggestions):
* Source/WebCore/html/TextFieldInputType.h:
* Source/WebCore/loader/EmptyClients.cpp:
(WebCore::EmptyChromeClient::createDataListSuggestionPicker):
* Source/WebCore/loader/EmptyClients.h:
* Source/WebCore/page/Chrome.cpp:
(WebCore::Chrome::createDataListSuggestionPicker):
* Source/WebCore/page/Chrome.h:
* Source/WebCore/page/ChromeClient.h:
* Source/WebCore/platform/DataListSuggestionPicker.h:
(WebCore::DataListSuggestionPicker::detach):
* Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.cpp:
(WebKit::WebChromeClient::createDataListSuggestionPicker):
* Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.h:
* Source/WebKit/WebProcess/WebCoreSupport/WebDataListSuggestionPicker.cpp:
(WebKit::WebDataListSuggestionPicker::handleKeydownWithIdentifier):
(WebKit::WebDataListSuggestionPicker::didSelectOption):
(WebKit::WebDataListSuggestionPicker::didCloseSuggestions):
(WebKit::WebDataListSuggestionPicker::close):
(WebKit::WebDataListSuggestionPicker::displayWithActivationType):
(WebKit::WebDataListSuggestionPicker::detach):
* Source/WebKit/WebProcess/WebCoreSupport/WebDataListSuggestionPicker.h:
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::didSelectDataListOption):
(WebKit::WebPage::didCloseSuggestions):
* Source/WebKitLegacy/mac/WebCoreSupport/WebChromeClient.h:
* Source/WebKitLegacy/mac/WebCoreSupport/WebChromeClient.mm:
(WebChromeClient::createDataListSuggestionPicker):

Canonical link: <a href="https://commits.webkit.org/285306@main">https://commits.webkit.org/285306@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/df9f67ca9e477343f3c814c4c191d68d6da6be80

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/72133 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/51554 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/24921 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/76298 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/23342 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/74248 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/59358 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/23164 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/56882 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/15395 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/75200 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/46730 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/62124 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/37322 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/43391 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/19598 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/21692 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/65297 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/19958 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/77978 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/16374 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/19129 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/65358 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/16421 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/62147 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/64617 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/12822 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/6468 "Passed tests") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/11081 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/47352 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/2136 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/48421 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/49709 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/48164 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebKit/WebProcess/WebCoreSupport/WebDataListSuggestionPicker.h:
```diff
@@ -48,21 +40,31 @@ namespace WebKit {
 
-    CheckedRef<WebCore::DataListSuggestionsClient> m_client;
-    WeakRef<WebPage> m_page;
+    CheckedPtr<WebCore::DataListSuggestionsClient> m_client;
```
This is not correct. Nothing is nulling out `m_client`. Since WebDataListSuggestionPicker is ref-counted, it can now theoretically outlive the client, in which case the CheckedPtr will cause a CRASH() if m_client is not nulled out before the client gets destroyed.

2 ways to fix this:
1. Use a WeakPtr instead of a CheckedPtr
2. Null out m_client when the client gets destroyed.

RupinMittal replied:
This should be fixed now. I just added a `detach()` function to null out `m_client` when it's being destroyed.

cdumez commented on Source/WebCore/html/TextFieldInputType.cpp:
```diff
@@ -999,8 +1001,11 @@ void TextFieldInputType::didSelectDataListOption(const String& selectedOption)
 void TextFieldInputType::didCloseSuggestions()
 {
-    m_cachedSuggestions = { };
+    if (RefPtr suggestionPicker = m_suggestionPicker)
```
I would write this as:
```
if (RefPtr suggestionPicker = std::exchange(m_suggestionPicker, nullptr))
    suggestionPicker->detach();
```

cdumez commented on Source/WebCore/html/TextFieldInputType.cpp:
```diff
@@ -999,8 +1001,11 @@ void TextFieldInputType::didSelectDataListOption(const String& selectedOption)
-    m_cachedSuggestions = { };
+    if (RefPtr suggestionPicker = m_suggestionPicker)
+        suggestionPicker->detach();
     m_suggestionPicker = nullptr;
```
Then you can drop this line.

cdumez commented on Source/WebCore/platform/DataListSuggestionPicker.h:
```diff
@@ -32,22 +32,17 @@
 public:
     virtual ~DataListSuggestionPicker() = default;
 
+    virtual void ref() const = 0;
```
Please submit `AbstractRefCounted`

RupinMittal committed:
[IsDeprecatedWeakRefSmartPointerException] Make WebDataListSuggestionPicker ref-counted
https://bugs.webkit.org/show_bug.cgi?id=281341
rdar://137771975

Reviewed by Chris Dumez.

Remove IsDeprecatedWeakRefSmartPointerException] by making WebDataListSuggestionPicker
ref-counted. Since WebDataListSuggestionPicker could now outlive WebPage or
DataListSuggestionsClient, we make those members WeakPtr and CheckedPtr and null-check
them before use. Accordingly, we ensure the DataListSuggestionsClient is set to nullptr
before the DataListSuggestionsClient is destroyed.

* Source/WebCore/html/TextFieldInputType.cpp:
(WebCore::TextFieldInputType::~TextFieldInputType):
(WebCore::TextFieldInputType::handleKeydownEvent):
(WebCore::TextFieldInputType::didCloseSuggestions):
(WebCore::TextFieldInputType::displaySuggestions):
(WebCore::TextFieldInputType::closeSuggestions):
* Source/WebCore/html/TextFieldInputType.h:
* Source/WebCore/loader/EmptyClients.cpp:
(WebCore::EmptyChromeClient::createDataListSuggestionPicker):
* Source/WebCore/loader/EmptyClients.h:
* Source/WebCore/page/Chrome.cpp:
(WebCore::Chrome::createDataListSuggestionPicker):
* Source/WebCore/page/Chrome.h:
* Source/WebCore/page/ChromeClient.h:
* Source/WebCore/platform/DataListSuggestionPicker.h:
(WebCore::DataListSuggestionPicker::detach):
* Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.cpp:
(WebKit::WebChromeClient::createDataListSuggestionPicker):
* Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.h:
* Source/WebKit/WebProcess/WebCoreSupport/WebDataListSuggestionPicker.cpp:
(WebKit::WebDataListSuggestionPicker::handleKeydownWithIdentifier):
(WebKit::WebDataListSuggestionPicker::didSelectOption):
(WebKit::WebDataListSuggestionPicker::didCloseSuggestions):
(WebKit::WebDataListSuggestionPicker::close):
(WebKit::WebDataListSuggestionPicker::displayWithActivationType):
(WebKit::WebDataListSuggestionPicker::detach):
* Source/WebKit/WebProcess/WebCoreSupport/WebDataListSuggestionPicker.h:
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::didSelectDataListOption):
(WebKit::WebPage::didCloseSuggestions):
* Source/WebKitLegacy/mac/WebCoreSupport/WebChromeClient.h:
* Source/WebKitLegacy/mac/WebCoreSupport/WebChromeClient.mm:
(WebChromeClient::createDataListSuggestionPicker):

Canonical link: https://commits.webkit.org/285306@main
